### PR TITLE
changelings

### PR DESCRIPTION
**Pull request by Dutch-VanDerLinde**
## About the PR
adds changeling antagonist, based on from tg station

current progress:
transformation system kinda complete, doesnt transfer bought shop abilities on transform which is something i gotta fix (and if the entity your transforming into gets gibbed it'll crash the game so gotta fix that too)

absorbing is sorta done, when you're done absorbing a guy you extract their DNA, replaces their blood with ferrochromic acid and gives them 200 genetic damage

you can cycle through the DNA you've stolen with an action

abilities that are done: arm blade, chitinous armor, dissonant shriek (emp blast), chameleon skin (invisibility)

## Why / Balance
cool
design doc: https://github.com/space-wizards/docs/pull/130
unfinished so draft

## Technical details
adds new method in Polymorph system, PolymorphEntityAsHumanoid which allows you to polymorph an entity with specific humanoid components

AddAction doesnt add the action if the entity already has that action prototype
Added new method in actionscontainersystem, HasAction. This lets you see if the given entity has that action prototype
Store listings dont show if the buyer has the ProductAction prototype

## Media
absorbing urist
image

cycling through dna strands
image

da evolution menu
image

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

### CODE REVIEW

**Comment by UbaserB** [comment]
YOOOO

**Comment by RedNiloc** [comment]
Holy, I love this. I hope it gets a horror form like in Goonstation.

**Comment by Jackal298** [comment]
I like it, I would also like to implement the Organic space suit, but I think this can be implemented in the future

**Comment by Brandon-Huu** [comment]
Sad

**Comment by deathride58** [comment]
You should probably make a design doc for this.

**Comment by Brandon-Huu** [comment]
> You should probably make a design doc for this.

>  need to make a design doc, working on that

**Comment by Dutch-VanDerLinde** [comment]
> You should probably make a design doc for this.

yep

**Comment by RedNiloc** [comment]
Will the drain victims look any different from just dead people?

**Comment by 2013HORSEMEATSCANDAL** [comment]
In 13 they'd get husked but we dont have husks yet so i dont know

**Comment by MrToofters** [comment]
This is really cool id love to see that in the game but im slightly concerned about the 200 genetic damage. I dont know if itd be nice remove someone from the game like that. Maybe just 100 would do?

**Comment by Yuoko** [comment]
> This is really cool id love to see that in the game but im slightly concerned about the 200 genetic damage. I dont know if itd be nice remove someone from the game like that. Maybe just 100 would do?

I think this good idea 💡

**Comment by Ilya246** [comment]
> This is really cool id love to see that in the game but im slightly concerned about the 200 genetic damage. I dont know if itd be nice remove someone from the game like that. Maybe just 100 would do?

lings are supposed to roundremove you

**Comment by BreadyToad** [comment]
> This is really cool id love to see that in the game but im slightly concerned about the 200 genetic damage. I dont know if itd be nice remove someone from the game like that. Maybe just 100 would do?

That's the intention here, because changelings are supposed to replace their targets. If only 100 genetic damage is done changeling players will still gib/dispose of the body to not be found out

**Comment by MrToofters** [comment]
> > This is really cool id love to see that in the game but im slightly concerned about the 200 genetic damage. I dont know if itd be nice remove someone from the game like that. Maybe just 100 would do?
> 
> That's the intention here, because changelings are supposed to replace their targets. If only 100 genetic damage is done changeling players will still gib/dispose of the body to not be found out

Fair Enough

**Comment by MrToofters** [comment]
also imay have missed it but how much chemicals do you get and do they regenerate?

**Comment by lzk228** [comment]
waiting for gazillion messages review from sloth :trollface:

**Comment by deltanedas** [comment]
in ss13 i think its possible to revive husks, so give 1000 burn or something

**Comment by K-Dynamic** [comment]
Off-topic but I've always found it weird lings don't infect the whole station like The Thing

**Comment by brainfood1183** [comment]
Ling is a classic antag, cannot wait to eventually see it merged. I thinkone of the main fears with the ling is that it should be a covert hidden type antagonist that is attempting to avoid detection and not an all out killing machine going toe to toe with security. Hopefully a decent balance can be found between both the usefulness of its abilities and them not being overtly overpowered in direct confrontation.

**Comment by Dutch-VanDerLinde** [comment]
> in ss13 i think its possible to revive husks, so give 1000 burn or something

i think when you get husked from a changeling in 13 it makes you unhuskable, only burn damage husks are revivable

**Comment by Dutch-VanDerLinde** [comment]
> also imay have missed it but how much chemicals do you get and do they regenerate?

0.5 chemicals a second, maximum of 75 can be increased through absorbing people

**Comment by keronshb** [comment]
I have a shop PR in the works that will give it action upgrade functionality. Once #24005 is merged, I'll be doing that PR soon after. Lings could benefit from this system.

File: Content.Server/Changeling/EntitySystems/ChangelingSystem.cs
Comment (deltanedas) [inline Content.Server/Changeling/EntitySystems/ChangelingSystem.cs:143]: put this in its own comp and share it with revenant :trollface:

**Review by deltanedas — CHANGES_REQUESTED** [review]

File: Content.Server/Changeling/EntitySystems/ChangelingSystem.cs
Comment (Dutch-VanDerLinde) [reply Content.Server/Changeling/EntitySystems/ChangelingSystem.cs:143]: i think this should be good now

**Comment by keronshb** [comment]
Added Frozen due to the antag freeze. Please still feel free to work on the doc & PR as needed. #8524

**Comment by Hmeister-fake** [comment]
> Added Frozen due to the antag freeze. Please still feel free to work on the doc & PR as needed. #8524

So ninja, thief and terminator don't apply to antag freeze and this and vampires does? Whats the difference?

**Comment by Emisse** [comment]
they were all under freeze

**Comment by Carolyn3114** [comment]
will this get any other abilities like head slug, tentacle, or lesser form?

**Comment by AJCM-git** [comment]
> > Added Frozen due to the antag freeze. Please still feel free to work on the doc & PR as needed. #8524
> 
> So ninja, thief and terminator don't apply to antag freeze and this and vampires does? Whats the difference?

A freeze doesn't necessarily mean it wont be merged (there is a tag for DO NOT MERGE), it just means there will be more scrutiny in the design and implementation. I imagine those prs didnt get the notice (tho they should have anyways) since the design had been agreed on already

**Comment by deltanedas** [comment]
thief and terminator had design docs made

ninja had doc of "le tg did it" which is why docs are required :trollface:

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Please don't turn ChangelingComponent into a godcomponent, use composition as it has 37 fields and like 3 abilities.

**Comment by metalgearsloth** [comment]
As abraham said antag PRs are scrutinisied significantly more due to the history of shitcode so expect a lot of reviews prior to merge.

**Comment by keronshb** [comment]
> > Added Frozen due to the antag freeze. Please still feel free to work on the doc & PR as needed. #8524
> 
> So ninja, thief and terminator don't apply to antag freeze and this and vampires does? Whats the difference?

image

**Comment by Hmeister-real** [comment]
> A freeze doesn't necessarily mean it wont be merged (there is a tag for DO NOT MERGE), it just means there will be more scrutiny in the design and implementation.

Ah didn't know this, thanks.

**Comment by Dutch-VanDerLinde** [comment]
> Please don't turn ChangelingComponent into a godcomponent, use composition as it has 37 fields and like 3 abilities.

What's composition

**Comment by RedNiloc** [comment]
God, I hope this actually ends up merging with the real game. I have been waiting a long time for ling to be added.

**Comment by xKREKSx** [comment]
I did a changeling resprite for space station 13 not too long ago.
but I never used it, do you want to take a look at it?

**Comment by Dutch-VanDerLinde** [comment]
> I did a changeling resprite for space station 13 not too long ago. but I never used it, do you want to take a look at it?

Sure

**Comment by xKREKSx** [comment]
> > I did a changeling resprite for space station 13 not too long ago. but I never used it, do you want to take a look at it?
> 
> Sure

image
image
image
Also the armblade and tentacle have a ripple animation when they're in your hand.

**Comment by Dutch-VanDerLinde** [comment]
> > > I did a changeling resprite for space station 13 not too long ago. but I never used it, do you want to take a look at it?
> > 
> > 
> > Sure
> 
> image image image Also the armblade and tentacle have a ripple animation when they're in your hand.

Nice

**Comment by UbaserB** [comment]
> > > I did a changeling resprite for space station 13 not too long ago. but I never used it, do you want to take a look at it?
> > 
> > 
> > Sure
> 
> Also the armblade and tentacle have a ripple animation when they're in your hand.

i actually really enjoy these sprites

**Comment by Dutch-VanDerLinde** [comment]
> > > > > I did a changeling resprite for space station 13 not too long ago. but I never used it, do you want to take a look at it?
> > > > 
> > > > 
> > > > Sure
> > > 
> > > 
> > > image image image Also the armblade and tentacle have a ripple animation when they're in your hand.
> > 
> > 
> > Nice
> 
> If you want I can drop them to you in dm on discord, just tell me in what format etc

sure, dm me at RICKY (discord username) in png

**Comment by Gielk0** [comment]
i really like this but why is it frozen?

**Comment by CrigCrag** [comment]
> i really like this but why is it frozen?

All new antag PRs are frozen due to the antag freeze. HOWEVER, a frozen PR does not mean it won't be merged, it simply will be held to higher quality standards.

**Comment by RedNiloc** [comment]
How many other changelings will there be in one round?

**Comment by Dutch-VanDerLinde** [comment]
> How many other changelings will there be in one round?

8 at maximum

**Comment by RedNiloc** [comment]
You still working on this?

**Comment by Catdere** [comment]
ling in maints 🗣️🗣️

**Comment by Fazansen** [comment]
I hope he's still working on this antagonist.😔

**Comment by RedNiloc** [comment]
> I hope he's still working on this antagonist.😔

Ither he is taking a long break or he is done with it (given up).

**Comment by RedNiloc** [comment]
yo, lets go. He started workin on it again.

**Comment by Ammoniacres** [comment]
Any way people can help with this?

**Comment by Emisse** [comment]
Closed due to feature freeze May 10th-June 14th. Comment to have it reopen after this.

**Comment by albassort** [comment]
This would be cool to reopen?

**Pull request closed without merging**
